### ISSUE TITLE
`RandomCoin` as a trait

### ISSUE THREAD

**Issue by Al-Kindi-0** [facebook/winterfell#162]
This issue aims to start a discussion on the merits of changing `RandomCoin<B: StarkField, H: Hasher>` from a struct to a trait. The reasons for this being that for some choices of `H` (specifically the ones based on the sponge construction) as well as of `B` (e.g. when it is the base field of the algebraic hasher `H`) seeding (which can be done with absorbing) and drawing random values (which can be done by squeezing) can be achieved in more efficient ways.

**Comment by irakliyk** [comment]
This would be great! I've been thinking about something like this as well. Maybe the trait could look like this:
```Rust
pub trait RandomCoin {
    type BaseField: StarkField;
    type Hasher: Hasher;

    // REQUIRED METHODS
    // --------------------------------------------------------------------------------------------

    fn new(seed: &[Self::BaseField]) -> Self;

    fn reseed(&mut self, data: <Self::Hasher as Hasher>::Digest);

    fn reseed_with_int(&mut self, value: u64);

    fn check_leading_zeros(&self, value: u64) -> u32;

    fn leading_zeros(&self) -> u32;

    fn draw<E: FieldElement<BaseField = Self::BaseField>>(&mut self) -> Result<E, RandomCoinError>;

    // PROVIDED METHODS
    // --------------------------------------------------------------------------------------------

    fn draw_pair<E>(&mut self) -> Result<(E, E), RandomCoinError>
    where
        E: FieldElement<BaseField = Self::BaseField>,
    {
        Ok((self.draw()?, self.draw()?))
    }

    fn draw_triple<E>(&mut self) -> Result<(E, E, E), RandomCoinError>
    where
        E: FieldElement<BaseField = Self::BaseField>,
    {
        Ok((self.draw()?, self.draw()?, self.draw()?))
    }

    fn draw_integers(
        &mut self,
        num_values: usize,
        domain_size: usize,
    ) -> Result<Vec<usize>, RandomCoinError> {
        todo!()
    }
}
```

**Comment by irakliyk** [comment]
Closed by #176.
